lib/ain-evm/src/core.rs

```diff
@@ -727,6 +729,113 @@ impl EVMCoreService {
+            block_number, state_root
+        );
+
+        let mut vicinity = Vicinity::from(block_header);
+        vicinity.origin = caller;
+        vicinity.gas_price = arguments.gas_price;
+
+        let backend = EVMBackend::from_root(
+            state_root,
+            Arc::clone(&self.trie_store),
+            Arc::clone(&self.storage),
+            vicinity,
+            None,
+        )
+        .map_err(|e| format_err!("Could not restore backend {}", e))?;
+
+        static CONFIG: Config = AinExecutor::CONFIG;
+        let metadata = StackSubstateMetadata::new(gas_limit, &CONFIG);
+        let state = MemoryStackState::new(metadata.clone(), &backend);
+        let al_state = MemoryStackState::new(metadata, &backend);
+        let precompiles = MetachainPrecompiles;
+        let mut executor = StackExecutor::new_with_precompiles(state, &CONFIG, &precompiles);
+        let mut al_executor = StackExecutor::new_with_precompiles(al_state, &CONFIG, &precompiles);
+
+        let mut listener = crate::eventlistener::StorageAccessListener::new();
+
+        runtime_using(&mut listener, move || {
+            let access_list = access_list
+                .into_iter()
+                .map(|x| (x.address, x.storage_keys))
+                .collect::<Vec<_>>();
+
+            match to {
+                Some(to) => {
+                    executor.transact_call(
+                        caller,
+                        to,
+                        value,
+                        data.to_vec(),
+                        gas_limit,
+                        access_list,
+                    );
+                }
+                None => {
+                    executor.transact_create(caller, value, data.to_vec(), gas_limit, access_list);
+                }
+            };
+        });
+
+        let al: AccessList = listener
+            .access_list
+            .into_iter()
+            .map(|(address, storage_keys)| AccessListItem {
+                address,
+                storage_keys: Vec::from_iter(storage_keys),
+            })
+            .collect();
+
+        match to {
+            Some(to) => {
+                al_executor.transact_call(
+                    caller,
+                    to,
+                    value,
+                    data.to_vec(),
+                    gas_limit,
+                    al.iter()
+                        .map(|item| (item.address, item.storage_keys.clone()))
+                        .collect(),
+                );
+            }
+            None => {
+                al_executor.transact_create(
+                    caller,
+                    value,
+                    data.to_vec(),
+                    gas_limit,
+                    al.iter()
+                        .map(|item| (item.address, item.storage_keys.clone()))
+                        .collect(),
+                );
+            }
+        };
+
+        Ok((al, al_executor.used_gas()))
+    }
```

Comment:
Why is there a need to re-execute the tx on the executor instance again?

Reply:
From what I am seeing, the results are never propagated either way from this execution nor does it change the access list obtained from the intial runtime listener that accumulates the access list.

Reply:
Ah, sorry cancel what I am saying. It is to obtain the execution gas usage with the access list populated.